File: src/engine/modules/command.ts
// src/engine/modules/command.ts
interface CommandCheckOptions {
  run: string;
  exit_code?: number;
  output_matches?: string;
}

interface CommandCheckResult {
  pass: boolean;
  reason?: string;
}

export async function runCommandCheck(
  options: CommandCheckOptions,
  cwd?: string
): Promise<CommandCheckResult> {
  const proc = Bun.spawnSync(["sh", "-c", options.run], {
    cwd,
    stderr: "pipe",
  });

  const expectedExit = options.exit_code ?? 0;
  if (proc.exitCode !== expectedExit) {
    return {
      pass: false,
      reason: `exit code ${proc.exitCode} (expected ${expectedExit})`,
    };
  }

  if (options.output_matches) {
    const stdout = new TextDecoder().decode(proc.stdout);
    if (!new RegExp(options.output_matches).test(stdout)) {
      return {
        pass: false,
        reason: `output did not match /${options.output_matches}/`,
      };
    }
  }

  return { pass: true };
}

File: internal/modules/command.go
package modules

import (
	"fmt"
	"os/exec"
	"regexp"
	"strings"

	"github.com/ConspiracyOS/contracts/internal/types"
)

// Result is an alias for types.Result for package-local convenience.
type Result = types.Result

// RunCommand executes a shell command and checks its exit code and optional output.
func RunCommand(c *types.CommandCheck, cwd string) Result {
	cmd := exec.Command("sh", "-c", c.Run)
	cmd.Dir = cwd
	out, err := cmd.Output()
	stdout := strings.TrimSpace(string(out))

	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			return Result{false, fmt.Sprintf("exec error: %v", err), ""}
		}
	}

	expected := c.ExitCode // default 0
	if exitCode != expected {
		return Result{false, fmt.Sprintf("exit code %d (expected %d)", exitCode, expected), stdout}
	}

	if c.OutputMatches != "" {
		re, err := regexp.Compile(c.OutputMatches)
		if err != nil {
			return Result{false, fmt.Sprintf("invalid output_matches regex: %v", err), stdout}
		}
		if !re.MatchString(stdout) {
			return Result{false, fmt.Sprintf("output %q did not match /%s/", stdout, c.OutputMatches), stdout}
		}
	}

	return Result{Pass: true, Evidence: stdout}
}
